clang/lib/Sema/SemaType.cpp

```diff
@@ -4887,9 +4887,18 @@ static TypeSourceInfo *GetFullTypeForDeclarator(TypeProcessingState &state,
                       cast<AutoType>(T)->getKeyword() !=
                           AutoTypeKeyword::Auto ||
                       cast<AutoType>(T)->isConstrained())) {
-            S.Diag(D.getDeclSpec().getTypeSpecTypeLoc(),
-                   diag::err_trailing_return_without_auto)
-                << T << D.getDeclSpec().getSourceRange();
+            SourceLocation Loc = D.getDeclSpec().getTypeSpecTypeLoc();
+            SourceRange SR = D.getDeclSpec().getSourceRange();
+            if (Loc.isInvalid()) {
+              TypeSourceInfo *TSI = nullptr;
+              S.GetTypeFromParser(FTI.getTrailingReturnType(), &TSI);
+              if (TSI) {
+                TypeLoc TSILoc = TSI->getTypeLoc();
+                Loc = TSILoc.getBeginLoc();
+                SR = TSILoc.getSourceRange();
+              }
+            }
+            S.Diag(Loc, diag::err_trailing_return_without_auto) << T << SR;
```

Comment:
I wonder if the diagnostic message looks better if we use `D.getDeclSpec().getSourceRange();` as the source range and `FTI.getTrailingReturnTypeLoc()` as the location?

Reply:
@Fznamznon thanks for the feedback! I think it could use the full range of the string, though I’m unsure if that’s acceptable...

```
template <typename T> struct B {
  // CHECK-NEXT: {{^}}  template <class U> B(U) -> B<int>;
  // CHECK-NEXT: {{^}}  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  template <class U> B(U) -> B<int>;
};
```

Reply:
IMO that looks fine.

Reply:
To confirm, should it attach to `B` and use the range of the _trailing return type_? I followed the GCC approach, which was the pointed in the issue - 
https://godbolt.org/z/esz9K5M95.

Reply:
Well, the diagnostic complains about missing `auto` in in the function declaration. There is no problem with `B` itself, so I think it is fine to use the full range. We don't have to follow GCC in all the aspects. 
cc @AaronBallman @Sirraide , WDYT?

Reply:
> There is no problem with `B` itself, so I think it is fine to use the full range.

It’s probably fine yeah. That might end up looking a bit chaotic if the function parameter list has e.g. 10 parameters and spans multiple lines, but you’d run into the same problem w/ a really long trailing return type. I don’t really have a strong opinion on this. Either option is better than including no source range at all ;Þ

Reply:
I guess another option would be to put the caret before the function name, but I don’t know how well that would work; i.e. something like this:
```
1 |    foo() -> bar;
  |   ^
```

Reply:
@Fznamznon @Sirraide Is the following range acceptable?

```cpp
template <typename T> struct B {
  // CHECK:      error: function with trailing return type must specify return type 'auto', not 'void'
  // CHECK-NEXT: {{^}}  template <class U> B(U) -> B<int>;
  // CHECK-NEXT: {{^}}                     ~~~~~~~~^~~~~~{{$}}
  template <class U> B(U) -> B<int>;
};
```